plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -149,24 +153,27 @@ private fun AnnotationVisitor.visitKotlinMetadata(header: Metadata) {
     visitEnd()
 }
 
-private fun KmClass.removePrivateDeclarations(removeCopyAlongWithConstructor: Boolean) {
-    constructors.removeIf { it.visibility.isPrivate }
+private fun KmClass.removePrivateDeclarations(removeCopyAlongWithConstructor: Boolean, classesToBeDeleted: Set<String>, pruneClass: Boolean) {
+    constructors.removeIf { pruneClass || it.visibility.isPrivate }
     (this as KmDeclarationContainer).removePrivateDeclarations(
-        copyFunShouldBeDeleted(removeCopyAlongWithConstructor)
+        copyFunShouldBeDeleted(removeCopyAlongWithConstructor),
+        pruneClass,
     )
+    nestedClasses.removeIf { "$name\$$it" in classesToBeDeleted }
```

Comment:
Not sure if we need to be smarter here for pruned classes. Probably not, since all nested classes of a pruned should be considered effectively private.